### PR TITLE
Preliminary DeviceMesh + native c10d functional integration

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118498
* #118437
* __->__ #118423

### Summary
- Added `group_name` as the third field in `dim_group_infos`.
- `DeviceMeshTest` now runs both w/ and w/0 `_USE_NATIVE_C10D_FUNCTIONAL=1` in CI.

### Other fixes
- Convert `reduceOp` to lower case before passing it into c10d_functional ops.
- Added a finalizer to handle unwaited collectives (this mirrors the treatment for Python functional collective ops).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -254,9 +254,13 @@ def _get_or_create_default_group(self):
...
-            # group tag/ranks associated with each mesh dimension, each mesh dimension should
-            # have one sub-group per rank
-            dim_group_infos: List[Tuple[str, List[int]]] = []
+            # tag/ranks/group_name associated with each mesh dimension, each
+            # mesh dimension should have one sub-group per rank
+            #
+            # TODO(yifu): remove tag and ranks once we fully migrate to native
+            # functional collectives. See details in:
+            # https://github.com/pytorch/pytorch/issues/93173#issuecomment-1907095208
+            dim_group_infos: List[Tuple[str, List[int], str]] = []
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:263]: quick comment: would `tag` have the same semantic with `group_name`? i.e. are they the same str? If they are maybe we can reuse the tag?

Comment (yifuwang) [reply torch/distributed/device_mesh.py:263]: They are different things. tag isn't always present so it's tricky to reuse.

**Comment by wconstab** [comment]
Suggest you land the C++ FakePG first (restack for a different landing order) and then ensure that when you land the native_c10d functional ops, both 0 and 1 paths are both CI tested from first place

**Comment by yifuwang** [comment]
@wconstab reordered the PRs as you suggested.

**Review by wanchaol — APPROVED** [review]
Nice, sgtm!

**Review by LucasLLC — APPROVED** [review]
lgtm, tagging @wz337 since she is the device mesh expert

**Review by wconstab — APPROVED** [review]
cool lgtm! (maybe update PR desc, still says tests fail bc of fake PG)

**Pull request closed without merging**